tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/IModelBindingResult.cs

```diff
@@ -0,0 +1,37 @@
+// <copyright file="IModelBindingResult.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+#if !NETFRAMEWORK
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AspNetCore
+{
+    /// <summary>
+    /// model binding result duck type
+    /// </summary>
+    internal interface IModelBindingResult
```

Comment:
I think this should be a `[DuckCopy]` struct, as it's not used _directly_ in a method constraint, and doesn't call any methods

Reply:
The problem is if I add it as a DuckCopy and then as a Property of the `IDefaultModelBindingContet `like this 
![image](https://user-images.githubusercontent.com/8353486/158238534-49dace8f-9b75-47ed-9904-e5583bd215d4.png)
![image](https://user-images.githubusercontent.com/8353486/158238805-34790d2c-7e0b-4bf2-99ad-84268e3df7ac.png)

it's never duck typed correctly from the `IDefaultModelBindingContext`, I never get IsModelSet to true for example, but I get no error either... I tried many combinations and that was the only one that would work... but maybe I'm missing something?

Reply:
[DuckCopy] structs uses fields instead of properties for mapping, that's the reason it didn't work. You can find examples here: https://github.com/DataDog/dd-trace-dotnet/blob/master/tracer/src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs#L895-L927
But TLDR, it should work if you remove `{ get; }` and convert it to fields.

Reply:
oh thanks will try!!

Reply:
the problem is that I'm in an interface, the type constraint in the `methodEnd`, so I cant have fields.. 
I tried 
```
 [Duck(BindingFlags = DuckAttribute.DefaultFlags | System.Reflection.BindingFlags.IgnoreCase, Kind = DuckKind.Property)]
        ModelBindingResult Result { get; set; }
```
but it didnt work, so I guess I have to remove the interface constraint and try a duck cast within the method body?

Reply:
I tried 
```
 if (instance.TryDuckCast<DefaultModelBindingContext>(out var defaultModelBindingContext))
                {
```
and 
```
[DuckCopy]
    internal struct DefaultModelBindingContext
    {
        /// <summary>
        /// Gets or sets the Model
        /// </summary>
        internal object Model;

        /// <summary>
        /// Gets the HttpContext
        /// </summary>
        [Duck(BindingFlags = DuckAttribute.DefaultFlags | BindingFlags.IgnoreCase, Kind = DuckKind.Property)]
        internal HttpContext HttpContext;

        [Duck(BindingFlags = DuckAttribute.DefaultFlags | BindingFlags.IgnoreCase)]
        internal ModelBindingResult Result;

        [Duck(BindingFlags = DuckAttribute.DefaultFlags | BindingFlags.IgnoreCase)]
        internal bool IsTopLevelObject;

        [Duck(BindingFlags = DuckAttribute.DefaultFlags | BindingFlags.IgnoreCase)]
        internal BindingSource BindingSource;

        /// <summary>
        /// Gets or sets the ValueProvider
        /// </summary>
        internal object ValueProvider;
    }
```
but Result isn't set though, I get no error 🤔

Reply:
the real object is:

`public class DefaultModelBindingContext : ModelBindingContext
  {
    private IValueProvider _originalValueProvider;
    private ActionContext _actionContext;
    private ModelStateDictionary _modelState;
    private ValidationStateDictionary _validationState;
    private DefaultModelBindingContext.State _state;
    private readonly Stack<DefaultModelBindingContext.State> _stack = new Stack<DefaultModelBindingContext.State>();

    /// <inheritdoc />
    public override ActionContext ActionContext
    {
      get => this._actionContext;
      set => this._actionContext = value != null ? value : throw new ArgumentNullException(nameof (value));
    }

    /// <inheritdoc />
    public override string FieldName
    {
      get => this._state.FieldName;
      set => this._state.FieldName = value != null ? value : throw new ArgumentNullException(nameof (value));
    }

    /// <inheritdoc />
    public override object Model
    {
      get => this._state.Model;
      set => this._state.Model = value;
    }

    /// <inheritdoc />
    public override ModelMetadata ModelMetadata
    {
      get => this._state.ModelMetadata;
      set => this._state.ModelMetadata = value != null ? value : throw new ArgumentNullException(nameof (value));
    }

    /// <inheritdoc />
    public override string ModelName
    {
      get => this._state.ModelName;
      set => this._state.ModelName = value != null ? value : throw new ArgumentNullException(nameof (value));
    }

    /// <inheritdoc />
    public override ModelStateDictionary ModelState
    {
      get => this._modelState;
      set => this._modelState = value != null ? value : throw new ArgumentNullException(nameof (value));
    }

    /// <inheritdoc />
    public override string BinderModelName
    {
      get => this._state.BinderModelName;
      set => this._state.BinderModelName = value;
    }

    /// <inheritdoc />
    public override BindingSource BindingSource
    {
      get => this._state.BindingSource;
      set => this._state.BindingSource = value;
    }

    /// <inheritdoc />
    public override bool IsTopLevelObject
    {
      get => this._state.IsTopLevelObject;
      set => this._state.IsTopLevelObject = value;
    }

    /// <summary>
    /// Gets or sets the original value provider to be used when value providers are not filtered.
    /// </summary>
    public IValueProvider OriginalValueProvider
    {
      get => this._originalValueProvider;
      set => this._originalValueProvider = value != null ? value : throw new ArgumentNullException(nameof (value));
    }

    /// <inheritdoc />
    public override IValueProvider ValueProvider
    {
      get => this._state.ValueProvider;
      set => this._state.ValueProvider = value != null ? value : throw new ArgumentNullException(nameof (value));
    }

    /// <inheritdoc />
    public override Func<ModelMetadata, bool> PropertyFilter
    {
      get => this._state.PropertyFilter;
      set => this._state.PropertyFilter = value;
    }

    /// <inheritdoc />
    public override ValidationStateDictionary ValidationState
    {
      get => this._validationState;
      set => this._validationState = value != null ? value : throw new ArgumentNullException(nameof (value));
    }

    /// <inheritdoc />
    public override ModelBindingResult Result
    {
      get => this._state.Result;
      set => this._state.Result = value;
    }
    ....
`

Reply:
fixed in fixed in [3754fbf](https://github.com/DataDog/dd-trace-dotnet/pull/2495/commits/3754fbfcbb54a8c95ba4d1aadc0ad008dc74848e)